### PR TITLE
update for MS3 type of cleavable crosslink data.

### PR DESCRIPTION
**Pull request by lonelu**
Changes include:
1. Data loading bug
2. Child Scan match
3. Parent ions information for cleavable crosslinker.

### CODE REVIEW

**Comment by lonelu** [comment]
This pull request is ready. I have these weird data in \\bison\share\Users\Lei\XL_SyntheticLibrary. Let me know if you have any questions running the data for testing.

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 1]
```diff
@@ -189,6 +199,9 @@ protected override MetaMorpheusEngineResults RunSpecific()
             return new MetaMorpheusEngineResults(this);
         }
 
+        /// <summary>
+        /// 
+        /// </summary>
```
Comment (zrolfs) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: ?

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 2]
```diff
@@ -198,20 +211,33 @@ private void GenerateCrosslinkModifications(Crosslinker crosslinker)
...
-
+                
                 if (XLPrecusorSearchMode.Accepts(scan.PrecursorMass, massTable[id]) >= 0)
                 {
                     PeptideIndex[id].Fragment(CommonParameters.DissociationType, FragmentationTerminus.Both, products);
                     var matchedFragmentIons = MatchFragmentIons(scan, products, CommonParameters);
                     double score = CalculatePeptideScore(scan.TheScan, matchedFragmentIons);
 
-                    if (score > byteScoreCutoff)
+                    if (score > (double)byteScoreCutoff)
```
Comment (zrolfs) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: double is unnecessary

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 3]
```diff
@@ -293,13 +324,16 @@ private void FindCrosslinkedPeptide(Ms2ScanWithSpecificMass scan, List<int> idsO
...
+                rank++;
             }
         }
+        
 
         /// <summary>
         /// Localizes the crosslink position on the alpha and beta peptides
         /// </summary>
-        private CrosslinkSpectralMatch LocalizeCrosslinkSites(Ms2ScanWithSpecificMass theScan, int alphaIndex, int betaIndex, Crosslinker crosslinker)
+        private CrosslinkSpectralMatch LocalizeCrosslinkSites(Ms2ScanWithSpecificMass theScan, int alphaIndex, int betaIndex, Crosslinker crosslinker, double[] xs, int[] intensityRanks)
```
Comment (zrolfs) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: "xs" should have a more descriptive name. Unclear what the variable is used for.

File: TaskLayer/MetaMorpheusTask.cs [hunk 4]
```diff
@@ -227,42 +228,39 @@ public static IEnumerable<Ms2ScanWithSpecificMass> GetMs2Scans(MsDataFile myMSDa
...
-                                    {
-                                        ms3ChildScan.RefineSelectedMzAndIntensity(precursorSpectrum.MassSpectrum);
-                                    }
-                                    catch (MzLibException ex)
-                                    {
-                                        Warn("Could not get precursor ion for MS3 scan #" + ms3ChildScan.OneBasedScanNumber + "; " + ex.Message);
-                                        continue;
-                                    }
-
+                                    //In current situation, do we need to perform the following function.
```
Comment (zrolfs) [inline TaskLayer/MetaMorpheusTask.cs]: why? elaborate in comment

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 5]
```diff
@@ -189,6 +199,9 @@ protected override MetaMorpheusEngineResults RunSpecific()
             return new MetaMorpheusEngineResults(this);
         }
 
+        /// <summary>
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: pleae add explanation

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 6]
```diff
@@ -471,17 +569,10 @@ private CrosslinkSpectralMatch LocalizeCrosslinkSites(Ms2ScanWithSpecificMass th
...
-
-                    if (crosslinker.Cleavable)
-                    {
-                        //TODO: re-enable intensity ranks
-                        //psmCrossAlpha.ParentIonMaxIntensityRanks = psmCrossAlpha.MatchedFragmentIons.Where(p => p.NeutralTheoreticalProduct.ProductType == ProductType.M).Select(p => p.IntensityRank).ToList();
-                        //localizedAlpha.ParentIonMaxIntensityRanks = new List<int>();
-
-                        //localizedAlpha.ParentIonExistNum = psmCrossAlpha.ParentIonMaxIntensityRanks.Count;
-                    }
+                    localizedAlpha.XLTotalScore = localizedAlpha.Score + localizedBeta.Score;
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: redundant?

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 7]
```diff
@@ -515,14 +606,21 @@ private List<MatchedFragmentIon> ScoreChildScan(Ms2ScanWithSpecificMass parentSc
                     mainPeptide.OneBasedStartResidueInProtein, mainPeptide.OneBasedEndResidueInProtein,
                     mainPeptide.CleavageSpecificityForFdrCategory, mainPeptide.PeptideDescription,
                     mainPeptide.MissedCleavages, mod, mainPeptide.NumFixedMods);
-
-                peptideWithMod.Fragment(CommonParameters.ChildScanDissociationType, FragmentationTerminus.Both, childProducts);
+          
+                    peptideWithMod.Fragment(CommonParameters.MS3ChildScanDissociationType, FragmentationTerminus.Both, childProducts);              
             }
-            else if (Math.Abs(childScan.PrecursorMass - parentScan.PrecursorMass) < 0.01 && CommonParameters.DissociationType != CommonParameters.ChildScanDissociationType)
+            else if (Crosslinker.Cleavable && Math.Abs(childScan.PrecursorMass - parentScan.PrecursorMass) < 0.01)
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: why hard coded mass differnce

File: EngineLayer/CommonParameters.cs [hunk 8]
```diff
@@ -51,7 +51,8 @@ public CommonParameters()
             ListOfModsFixed = listOfModsFixed ?? new List<(string, string)> { ("Common Fixed", "Carbamidomethyl on C"), ("Common Fixed", "Carbamidomethyl on U") };
             DissociationType = dissociationType;
             SeparationType = separationType;
-            ChildScanDissociationType = childScanDissociationType;
+            MS2ChildScanDissociationType = ms2childScanDissociationType;
```
Comment (trishorts) [inline EngineLayer/CommonParameters.cs]: all these dissociation types are getting to be a mess. maybe we should think about a less messy solution. Perhaps a collection of dissociation types?

File: EngineLayer/CommonParameters.cs [hunk 9]
```diff
@@ -103,9 +104,11 @@ public CommonParameters()
         public int MaxHeterozygousVariants { get; private set; }
         public int MinVariantDepth { get; private set; }
         public DissociationType DissociationType { get; private set; }
-        public DissociationType ChildScanDissociationType { get; private set; }
         public string SeparationType { get; private set; }
 
+        public DissociationType MS2ChildScanDissociationType { get; private set; }
```
Comment (trishorts) [inline EngineLayer/CommonParameters.cs]: same

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 10]
```diff
@@ -314,7 +348,7 @@ private CrosslinkSpectralMatch LocalizeCrosslinkSites(Ms2ScanWithSpecificMass th
             }
             else
             {
-                List<int> possibleAlphaXlSites = CrosslinkSpectralMatch.GetPossibleCrosslinkerModSites(crosslinker.CrosslinkerModSites.ToCharArray(), PeptideIndex[alphaIndex], CommonParameters.DigestionParams.InitiatorMethionineBehavior, CleaveAtCrosslinkSite);
+                List<int> possibleAlphaXlSites =CrosslinkSpectralMatch.GetPossibleCrosslinkerModSites(crosslinker.CrosslinkerModSites.ToCharArray(), PeptideIndex[alphaIndex], CommonParameters.DigestionParams.InitiatorMethionineBehavior, CleaveAtCrosslinkSite);
                 List<int> possibleBetaXlSites = CrosslinkSpectralMatch.GetPossibleCrosslinkerModSites(crosslinker.CrosslinkerModSites2.ToCharArray(), PeptideIndex[betaIndex], CommonParameters.DigestionParams.InitiatorMethionineBehavior, CleaveAtCrosslinkSite);
 
                 pairs.Add(new Tuple<List<int>, List<int>>(possibleAlphaXlSites, possibleBetaXlSites));
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: what are "pairs" alpha/beta?

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 11]
```diff
@@ -515,14 +606,21 @@ private List<MatchedFragmentIon> ScoreChildScan(Ms2ScanWithSpecificMass parentSc
...
-                childProducts = CrosslinkedPeptide.XlGetTheoreticalFragments(CommonParameters.ChildScanDissociationType,
-                    Crosslinker, new List<int> { possibleSite }, otherPeptide.MonoisotopicMass, mainPeptide).First().Item2;
+                if (childScan.TheScan.MsnOrder == 2)
+                {
+                    childProducts = CrosslinkedPeptide.XlGetTheoreticalFragments(CommonParameters.MS2ChildScanDissociationType,
+                        Crosslinker, new List<int> { possibleSite }, otherPeptide.MonoisotopicMass, mainPeptide).First().Item2;
+                }
+                else
+                {
+                    childProducts = CrosslinkedPeptide.XlGetTheoreticalFragments(CommonParameters.MS3ChildScanDissociationType,
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: unit test?

**Comment by trishorts** [comment]
much of this PR is covered by unit test. So that is looking good. maybe just a couple spots missing/

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 7 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: For example, for HCD-ETD type of data, it is better to make sure they have the same precursor mass.

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 10 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: The crosslink can crosslink same or different amino acid. Pairs are potential crosslink sites for alpha or beta.

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 11 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: This is for a rare weird case. I prefer not to test it.
It tried to corver the situation that the MS3 scan have the same precursor mass as the MS2.  
This is so rare that I only see it in one data. In order to test it, I need to add the data in MM and just make sure the code run it through.

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 2 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: deleted

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 1 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: Comment added.

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 3 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: Name changed.

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs (hunk 6 above)
Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: deleted.

File: EngineLayer/CommonParameters.cs (hunk 8 above)
Comment (lonelu) [reply EngineLayer/CommonParameters.cs]: This covers all the situations. I had a disscussion with Rob today. We think it is OK to keep it like this right now.

Comment (trishorts) [reply EngineLayer/CommonParameters.cs]: we need a long term solution to the problem of keeping track of what dissociation types are where.  see: https://github.com/smith-chem-wisc/MetaMorpheus/issues/1834

**Pull request merged**